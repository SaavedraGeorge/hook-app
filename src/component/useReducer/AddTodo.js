import React from 'react';
import { useForm } from '../../hooks/useForm';

export const AddTodo = ({handleAddTodo}) => {
    const [{description}, handleInputChange, reset] = useForm({ 
        description: ''
     });
    
    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1){
            return
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false 
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar To Do</h4>
            <hr />
            <form onSubmit={ handleSubmit }>
                <input 
                    type="text" 
                    name="description" 
                    placeholder="add To Do" 
                    autoComplete='off'
                    className='form-control'
                    value={ description }
                    onChange={ handleInputChange }
                />
                <button 
                    className='btn btn-outline-primary col-12 mt-1'
                    type='submit'
                >Agregar</button>
            </form>
        </>
    );
};
