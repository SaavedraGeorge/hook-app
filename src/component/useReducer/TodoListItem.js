import React from 'react';

export const TodoListItem = ({todo, i, handleDelete, toggle }) => {
  
  return (
    <li
    className='list-group-item'
    key={todo.id}
    >
        <p 
            onClick={ () => toggle(todo.id)}
            className={`${ todo.done ? 'complete': ''}`}
        >
            {i+1}. { todo.desc }
        </p>
        <button 
            className='btn btn-danger'
            onClick={ () => handleDelete(todo.id) }
        >Eliminar</button>
    </li>
  );
};
