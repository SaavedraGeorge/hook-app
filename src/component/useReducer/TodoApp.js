import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';


import './style.css'
import { TodoList } from './TodoList';
import { AddTodo } from './AddTodo';

const init = () =>{
    
    // llama lo que hay en el loca storrage y si es vacio devuelve []
    return JSON.parse(localStorage.getItem('todos')) || [];
    
} 

export const TodoApp = () => {

    const [todos, dispatch] = useReducer( todoReducer, [], init );

    
    
    
     useEffect(() => {
       localStorage.setItem('todos', JSON.stringify(todos));
     
     }, [todos]);
     

    const handleAddTodo = ( newTodo ) => {
        
        dispatch({
            type: 'add',
            payload: newTodo
        });

    }

    const handleDelete = (idItem) => {
        
        const action ={
            type: 'delete',
            payload: idItem
        }

        dispatch(action);

    }

    const toggle = (idItem) => {
        dispatch( {
            type: 'toggle',
            payload: idItem 
        } )
    }

    return (
        <div>
            <h1>To Do App <small>{todos.length}</small></h1>
            <hr />

  

            <div className='row'>
                <div className='col-7'>
                    <TodoList 
                        todos = {todos}
                        handleDelete = {handleDelete}
                        toggle = {toggle}
                    /> 
                </div>
                <div className='col-5'>
                    <AddTodo handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}
