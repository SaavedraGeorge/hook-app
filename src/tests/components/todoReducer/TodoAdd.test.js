import React from 'react';
import { shallow } from 'enzyme';
import { AddTodo } from '../../../component/useReducer/AddTodo';


describe('Pruebas en <AddTodo />', () => {

    const handleAddTodo = jest.fn();

    const wrapper = shallow( 
        <AddTodo 
            handleAddTodo={ handleAddTodo }
        /> 
    )


    test('debe de mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('NO debe de llamar handleAddTodo', () => {
        
        //busca el onSubmit del form
        const formSubmit = wrapper.find('form').prop('onSubmit');

        //ejecuta la accion onSubmit y manda el default para que corra
        formSubmit({ preventDefault(){} });

        //se espera que no se ejecute la accion, ya que no pasa la validacion
        expect( handleAddTodo ).toHaveBeenCalledTimes(0);


    });

    test('debe de llamar la funciÃ³n handleAddTodo', () => {
        
        //creo un nuevo value para agregar en mi simulacion
        const value = 'Aprender Firestore';
        //siulo el onchange y entrelo mi value en el input
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        });

        //se busca y se ejecuta el onsubmit con los cambios anterios al input
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} });

        //cuento cuantas veces el submit ah sido llamado, este vez es uno porque pasa la validacion
        expect( handleAddTodo ).toHaveBeenCalledTimes(1);

        //el handAddTodo es llamado con un objeto, pero esto puede entregar uno vacio
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) ); // { }
        //aca le pido la validacion con un objeto con datos, donde en el id le digo que es un numero (cualquiera)
        //el que recibe
        expect( handleAddTodo ).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        }); 

        //aca evaluo si despues de ejecutar el handleAddTodo se ejecuto el reset, deberia estar el input vacio
        expect( wrapper.find('input').prop('value') ).toBe('');




    })
    
    

    

    
})
