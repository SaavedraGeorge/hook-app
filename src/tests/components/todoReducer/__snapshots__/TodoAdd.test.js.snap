// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddTodo /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar To Do
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="add To Do"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary col-12 mt-1"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
